src/main/java/dev/morling/onebrc/CalculateAverage_jhonesto.java

```diff
@@ -0,0 +1,395 @@
+
+            max = max < t ? t : max;
+            min = min > t ? t : min;
+            sum += t;
+            count += ONE;
+
+        }
+
+        @Override
+        public String toString() {
+
+            return String.format("%s=%.1f/%.1f/%.1f",
+                    new String(station, StandardCharsets.UTF_8).trim(),
+                    Math.round(min) / TEN_DOUBLE,
+                    Math.round(sum / (double) count) / TEN_DOUBLE,
+                    Math.round(max) / TEN_DOUBLE);
+        }
+
+        public int compareTo(Measurement other) {
+            return new String(this.station).compareTo(new String(other.station));
+        }
+
+    }
+
+    // +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    // ++++++++++++++++++++++++++++ UTILS ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+    // +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+    // Get the end of partitions; b -> buffer, mbs -> max_buffer_size
+    private static long[] getEOP(int[] b, int mbs, File file) {
+
+        long eof[] = new long[b.length];
+
+        try (RandomAccessFile raf = new RandomAccessFile(file, READ)) {
+
+            for (int i = ZERO; i < b.length; i++) {
+
+                long size = ((long) mbs * i) +
+                        (long) (i == b.length - ONE ? b[i] : mbs);
+
+                raf.seek((size - ONE));
+
+                while (raf.read() != LINE_FEED) {
+                    raf.seek(--size - ONE);
+                }
+
+                eof[i] = size;
+            }
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        return eof;
+    }
+
+    public static byte[] asByteArrayCopy(byte[] arr, int length) {
+        byte[] b = new byte[length];
+        System.arraycopy(arr, ZERO, b, ZERO, length);
+        return b;
+    }
+
+    private static void resolveMeasurement(byte[] measurement, HashMap<Integer, Measurement> table) {
+
+        short temperature = 0;
+
+        int j = measurement.length - ONE;
+
+        for (short exp = ZERO; j > ZERO; j -= ONE, exp += ONE) {
+
+            if (measurement[j] == SEMICOLON) {
+                break;
+            }
+
+            if (measurement[j] == MINUS) {
+                temperature *= -ONE;
+            }
+            else if (measurement[j] == PERIOD) {
+                exp -= ONE;
+            }
+            else {
+                temperature += (short) ((measurement[j] % MOD_NUMBER) * pow(RADIX, exp));
+            }
+
+        }
+
+        byte[] station = asByteArrayCopy(measurement, j);
+
+        int hash = getHash(station);
+
+        Measurement m = table.get(hash);
+
+        if (m != null) {
+            m.setTemperature(temperature);
+
+        }
+        else {
+            m = new Measurement(station, temperature);
+            table.put(hash, m);
```

Comment:
Just using an integer hash as a key is prone to collisions. You can't reduce a 100 byte string to a 4 byte number without losing information.

Reply:
There is no test case for this, if you reproduce a collision, I will think about it.

Reply:
I can't create test cases for all potential hashing algorithms people may use. But it's clear that the entry is prone to hash collisions as-is and thus isn't valid. Let me know whether you'd like to rework it or retract it. Thx.

Reply:
1st -> I can't fix it if it's not broken.
2nd -> I'm using the Arrays.hashCode Algorithm used in the majority of implementations of this chalenge.
3rd -> A test case for collisions make more sense than a 1 row test.

There's two kind of situation here:

a - There's a collision "as-is". Source.: Voices of my mind.
b - There's a collision with stationA and stationB. 

Since my implementation passed all the tests, I will only fix situation B.

Thanks.

Reply:
>  There's a collision "as-is". Source.: Voices of my mind

I don't appreciate your tone here. It's a standard procedure to point out issues during a code review, no test is needed when they are that apparent. It's surprising that we even debate this; if you could unambigiously express the information from 100 random bytes using four bytes, you could compress everything indefinitely.

> I'm using the Arrays.hashCode Algorithm used in the majority of implementations of this chalenge.

They either use keys which implement identity based on the actual station name (not just their hash), or they implement custom maps with open addressing, checking station name for equality in case of hash collisions.

> A test case for collisions make more sense than a 1 row test.

Again, how would that test look like? People use different functions and seed values so it's not doable on a general level. I might add sum fuzzy testing, but it would come at the cost of taking things even longer to run. Anyways, here's a patch you can apply which adds a test that fails with your implementation:

```
---
 src/test/resources/samples/measurements-hashcollision.out | 1 +
 src/test/resources/samples/measurements-hashcollision.txt | 2 ++
 2 files changed, 3 insertions(+)
 create mode 100644 src/test/resources/samples/measurements-hashcollision.out
 create mode 100644 src/test/resources/samples/measurements-hashcollision.txt

diff --git a/src/test/resources/samples/measurements-hashcollision.out b/src/test/resources/samples/measurements-hashcollision.out
new file mode 100644
index 00000000..42af246f
--- /dev/null
+++ b/src/test/resources/samples/measurements-hashcollision.out
@@ -0,0 +1 @@
+{pjhvyxbwwuifdgwhwoxz=19.2/19.2/19.2, wsyrjwjcdpmqeypcjlwh=9.5/9.5/9.5}
diff --git a/src/test/resources/samples/measurements-hashcollision.txt b/src/test/resources/samples/measurements-hashcollision.txt
new file mode 100644
index 00000000..dd34715c
--- /dev/null
+++ b/src/test/resources/samples/measurements-hashcollision.txt
@@ -0,0 +1,2 @@
+pjhvyxbwwuifdgwhwoxz;19.2
+wsyrjwjcdpmqeypcjlwh;9.5
-- 
```

```
./test.sh jhonesto src/test/resources/samples/measurements-hashcollision.txt
Validating calculate_average_jhonesto.sh -- src/test/resources/samples/measurements-hashcollision.txt
1c1,2
< pjhvyxbwwuifdgwhwoxz;9.5;14.4;19.2
---
> pjhvyxbwwuifdgwhwoxz;19.2;19.2;19.2
> wsyrjwjcdpmqeypcjlwh;9.5;9.5;9.5
```

Reply:
I don't care if you appreciate my tone here. Think is cheap you need to prove your argument. 

Now I can treat this test accordingly. But, in my tests it's ok. There is no collision.

![hash-collision](https://github.com/gunnarmorling/1brc/assets/46565138/a5073577-70ec-4d1c-bfc3-6879ab4c1ac9)